artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java

```diff
@@ -3069,57 +3080,74 @@ private void move(final Transaction originalTX,
     * This method delivers the reference on the callers thread - this can give us better latency in the case there is nothing in the queue
     */
    private boolean deliverDirect(final MessageReference ref) {
-      synchronized (this) {
-         if (!supportsDirectDeliver) {
-            return false;
-         }
-         if (paused || !canDispatch() && redistributor == null) {
-            return false;
-         }
+      //The order to enter the deliverLock re QueueImpl::this lock is very important:
```

Comment:
rather than this, why not have the original method renamed and existing method with the lock delegating then to the renamed (but original) method. This way keeps the lock cleaner.

Reply:
I'm not sure, I don't want the original method to be called without using the proper lock order and having them in one place will help to avoid misusing: that's personal...
If not, we can split the method and add an assertion to ensure that the deliveryLock needs to be held by the current Thread too: it will avoid future misuses.

Reply:
Its private, its no different to someone calling deliver() not within the runner.